packages/transactions/src/new-signatures.ts

```diff
@@ -0,0 +1,96 @@
+import { Address, getAddressFromPublicKey } from '@solana/addresses';
+import { Decoder } from '@solana/codecs-core';
+import { getBase58Decoder } from '@solana/codecs-strings';
+import {
+    SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING,
+    SOLANA_ERROR__TRANSACTION__SIGNATURES_MISSING,
+    SolanaError,
+} from '@solana/errors';
+import { Signature, SignatureBytes, signBytes } from '@solana/keys';
+
+import { CompilableTransaction } from './compilable-transaction';
+import { compileMessage } from './message';
+import { OrderedMap } from './ordered-map';
+import { getCompiledMessageEncoder } from './serializers';
+
+export type TransactionMessageBytes = Uint8Array & { readonly __brand: unique symbol };
+
+export interface INewTransactionWithSignatures {
+    messageBytes: TransactionMessageBytes;
+    signatures: OrderedMap<Address, SignatureBytes | null>;
+}
+
+export interface INewFullySignedTransaction extends INewTransactionWithSignatures {
+    readonly __brand: unique symbol;
+}
+
+let base58Decoder: Decoder<string> | undefined;
+
+export function newGetSignatureFromTransaction(transaction: INewTransactionWithSignatures): Signature {
+    if (!base58Decoder) base58Decoder = getBase58Decoder();
+
+    // We have ordered signatures from the compiled message accounts
+    // first signature is the fee payer
+    const signatureBytes = transaction.signatures.firstValue();
+    if (!signatureBytes) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING);
+    }
+    const transactionSignature = base58Decoder.decode(signatureBytes);
+    return transactionSignature as Signature;
+}
+
+export async function newPartiallySignTransaction<TTransaction extends CompilableTransaction>(
+    keyPairs: CryptoKeyPair[],
+    transaction: TTransaction,
+): Promise<INewTransactionWithSignatures> {
+    const compiledMessage = compileMessage(transaction);
+    const transactionSigners = compiledMessage.staticAccounts.slice(0, compiledMessage.header.numSignerAccounts);
+    const messageBytes = getCompiledMessageEncoder().encode(compiledMessage) as TransactionMessageBytes;
+
+    const publicKeySignaturePairs = await Promise.all(
+        keyPairs.map(keyPair =>
+            Promise.all([getAddressFromPublicKey(keyPair.publicKey), signBytes(keyPair.privateKey, messageBytes)]),
+        ),
+    );
+    const publickKeySignatureMap = new Map(publicKeySignaturePairs);
+
+    const signatures = new OrderedMap<Address, SignatureBytes | null>();
+    for (const signerAddress of transactionSigners) {
+        const signature = publickKeySignatureMap.get(signerAddress) ?? null;
+        signatures.set(signerAddress, signature);
+    }
+
+    const out: INewTransactionWithSignatures = {
+        messageBytes,
+        signatures,
+    };
+    Object.freeze(out);
```

Comment:
That's an annoying limitation, but I'm inclined to just leave it as-is. The `CryptoKey` approach feels a bit overkill for this, and I think storing this as anything but a `Uint8Array` is just going to add extra copies whenever someone wants to do anything with it. Ultimately it's probably going to get serialised and sent somewhere, there are limits to how much we can help you if you start randomly modifying byte arrays!

Reply:
> Though I'm a bit confused why I'm not getting an error on the Object.freeze call…

It freezes shallowly. If we're not freezing every node in the tree elsewhere, we should be.

> …I think storing this as anything but a Uint8Array is just going to add extra copies whenever someone wants to do anything with it.

Copies are Bad™, but this very well may be literally the only place in `@solana/transactions` that you can't do `tx === otherTx` and know for sure that they're identical. Making `SignatureBytes` an `Array` would mean copying things back and forth, but we're also talking about fewer bytes than saying ‘Two coffees and a beer’ three times.

You're almost certainly correct, just wanted to give it another think.

Reply:
We had this conversation a little while ago on a PR and decided at the time to put a pin on it. Since we're breaking things here, maybe now is the time to revisit this?

https://github.com/solana-labs/solana-web3.js/pull/1710#discussion_r1389234589

Reply:
Alright cool this is using `ReadonlyUint8Array` now, added further up the stack